ssl/quic/quic_port.c

```diff
@@ -1329,13 +1347,78 @@ static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
         *scid = hdr->src_conn_id;
     }
 
+    /*
+     * Determine if we need to send a NEW_TOKEN frame
+     * If we validated a retry token, we should always
+     * send a NEW_TOKEN frame to the client
+     *
+     * If however, we validated a NEW_TOKEN, which may be
+     * reused multiple times, only send a NEW_TOKEN frame
+     * if the existing received token has 10% of its lifetime
+     * remaining.  This prevents us from constantly sending
+     * NEW_TOKEN frames on every connection when not needed
+     */
+    if (token.is_retry) {
+        *gen_new_token = 1;
+    } else {
+        if (time_diff < NEW_TOKEN_LIFETIME / 10)
+            *gen_new_token = 1;
+    }
+
     ret = 1;
 err:
     cleanup_validation_token(&token);
     OPENSSL_free(remote_addr);
     return ret;
 }
 
+static void generate_new_token(QUIC_CHANNEL *ch, BIO_ADDR *peer)
+{
+    QUIC_CONN_ID rscid = { 0 };
+    QUIC_VALIDATION_TOKEN token;
+    unsigned char buffer[ENCRYPTED_TOKEN_MAX_LEN];
+    unsigned char *ct_buf;
+    size_t ct_len;
+    size_t token_buf_len = 0;
+
+    /* Clients never send a NEW_TOKEN */
+    if (!ch->is_server)
+        return;
+
+    ct_buf = OPENSSL_zalloc(ENCRYPTED_TOKEN_MAX_LEN);
+    if (ct_buf == NULL)
+        return;
```

Comment:
Seems like a fatal error. Why is the return type from this function void?

Reply:
fatal is a strong word here.  Token generation happens at the port level (i.e. not strictly correlated to any specific SSL context in a using application, so raising an error on the error stack doesn't do much to clarify anything here).  Likewise, theres no failure to report to the peer if we don't send a new token - i.e. if we don't send a new token, then the "next" connection that the client establishes just has to go through the retry round trip.  So, its an error, but the only response to it is to not send a new token, which is sub-optimal, but ok, hence the blind return.